lisa/sut_orchestrator/libvirt/transformers.py

```diff
@@ -0,0 +1,523 @@
+
+class LibvirtInstallerTransformer(Transformer):
+    @classmethod
+    def type_name(cls) -> str:
+        return "libvirt_installer"
+
+    @classmethod
+    def type_schema(cls) -> Type[schema.TypedSchema]:
+        return InstallerTransformerSchema
+
+    @property
+    def _output_names(self) -> List[str]:
+        return []
+
+    def _internal_run(self) -> Dict[str, Any]:
+        runbook: InstallerTransformerSchema = self.runbook
+        assert runbook.connection, "connection must be defined."
+        assert runbook.installer, "installer must be defined."
+
+        node = quick_connect(runbook.connection, "libvirt_installer_node")
+
+        factory = subclasses.Factory[BaseInstaller](BaseInstaller)
+        installer = factory.create_by_runbook(
+            runbook=runbook.installer,
+            node=node,
+            log=self._log,
+        )
+        libvirt_version = installer.install()
+        self._log.info(f"installed libvirt version: {libvirt_version}")
+
+        if isinstance(node.os, Ubuntu):
+            node.execute("systemctl disable apparmor", shell=True, sudo=True)
+        node.execute("systemctl enable libvirtd", shell=True, sudo=True)
+        node.reboot()
+        return {}
+
+
+class QemuInstallerTransformer(Transformer):
+    @classmethod
+    def type_name(cls) -> str:
+        return "qemu_installer"
+
+    @classmethod
+    def type_schema(cls) -> Type[schema.TypedSchema]:
+        return InstallerTransformerSchema
+
+    @property
+    def _output_names(self) -> List[str]:
+        return []
+
+    def _internal_run(self) -> Dict[str, Any]:
+        runbook: InstallerTransformerSchema = self.runbook
+        assert runbook.connection, "connection must be defined."
+        assert runbook.installer, "installer must be defined."
+
+        node = quick_connect(runbook.connection, "qemu_installer_node")
+
+        factory = subclasses.Factory[BaseInstaller](BaseInstaller)
+        installer = factory.create_by_runbook(
+            runbook=runbook.installer,
+            node=node,
+            log=self._log,
+        )
+        qemu_version = installer.install()
+        self._log.info(f"installed qemu version: {qemu_version}")
+
+        node.reboot()
+        return {}
+
+
+class CloudHypervisorInstallerTransformer(Transformer):
+    @classmethod
+    def type_name(cls) -> str:
+        return "cloudhypervisor_installer"
+
+    @classmethod
+    def type_schema(cls) -> Type[schema.TypedSchema]:
+        return InstallerTransformerSchema
+
+    @property
+    def _output_names(self) -> List[str]:
+        return []
+
+    def _internal_run(self) -> Dict[str, Any]:
+        runbook: InstallerTransformerSchema = self.runbook
+        assert runbook.connection, "connection must be defined."
+        assert runbook.installer, "installer must be defined."
+
+        node = quick_connect(runbook.connection, "cloudhypervisor_installer_node")
+
+        factory = subclasses.Factory[BaseInstaller](BaseInstaller)
```

Comment:
That may be confused, which installers can work with this transformeron the same BaseInstaller. It can derive to two installers like "CloudHypervisorInstaller" and "QemuInstaller". And the transformer only accepts corresponding installers. The type_name can be simplified by this way.

Reply:
Thanks for suggesting this. Will apply it.